rust/lance-index/src/scalar/bloomfilter.rs

```diff
@@ -0,0 +1,1769 @@
+                    "BloomFilterIndex: 'fragment_id' column is not UInt64",
+                    location!(),
+                )
+            })?;
+
+        let block_start_col = data
+            .column_by_name("block_start")
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: missing 'block_start' column",
+                    location!(),
+                )
+            })?
+            .as_any()
+            .downcast_ref::<arrow_array::UInt64Array>()
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: 'block_start' column is not UInt64",
+                    location!(),
+                )
+            })?;
+
+        let block_size_col = data
+            .column_by_name("block_size")
+            .ok_or_else(|| {
+                Error::invalid_input("BloomFilterIndex: missing 'block_size' column", location!())
+            })?
+            .as_any()
+            .downcast_ref::<arrow_array::UInt64Array>()
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: 'block_size' column is not UInt64",
+                    location!(),
+                )
+            })?;
+
+        let bloom_filter_data_col = data
+            .column_by_name("bloom_filter_data")
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: missing 'bloom_filter_data' column",
+                    location!(),
+                )
+            })?
+            .as_any()
+            .downcast_ref::<arrow_array::BinaryArray>()
+            .ok_or_else(|| {
+                Error::invalid_input(
+                    "BloomFilterIndex: 'bloom_filter_data' column is not Binary",
+                    location!(),
+                )
+            })?;
+
+        let num_blocks = data.num_rows();
+        let mut blocks = Vec::with_capacity(num_blocks);
+
+        for i in 0..num_blocks {
+            let bloom_filter_data = if bloom_filter_data_col.is_valid(i) {
+                bloom_filter_data_col.value(i).to_vec()
+            } else {
+                Vec::new()
+            };
+
+            blocks.push(BloomFilterStatistics {
+                fragment_id: fragment_id_col.value(i),
+                block_start: block_start_col.value(i),
+                block_size: block_size_col.value(i) as usize,
+                bloom_filter_data,
+            });
+        }
+
+        Ok(Self {
+            blocks,
+            data_type,
+            number_of_items,
+            number_of_bytes,
+            store,
+            fri,
+            index_cache,
+        })
+    }
+
+    fn evaluate_block_against_query(
+        &self,
+        block: &BloomFilterStatistics,
+        query: &SargableQuery,
+    ) -> Result<bool> {
+        // If no bloom filter data is available, be conservative and return true
+        if block.bloom_filter_data.is_empty() {
```

Comment:
why could a block have no data? If there is no data the block shouldn't really exist I think